ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java

```diff
@@ -284,25 +303,39 @@ record = segment.getLogRecord(index);
       if (record == null) {
         return null;
       }
-      final LogEntryProto entry = segment.getEntryFromCache(record.getTermIndex());
+      final ReferenceCountedObject<LogEntryProto> entry = segment.getEntryFromCache(record.getTermIndex());
       if (entry != null) {
         getRaftLogMetrics().onRaftLogCacheHit();
+        entry.retain();
         return entry;
       }
     }
 
     // the entry is not in the segment's cache. Load the cache without holding the lock.
     getRaftLogMetrics().onRaftLogCacheMiss();
     cacheEviction.signal();
-    return segment.loadCache(record);
+    ReferenceCountedObject<LogEntryProto> loaded = segment.loadCache(record);
+    if (loaded != null) {
+      loaded.retain();
+    }
+    return loaded;
```

Comment:
We should revert this part, i.e.
```java
@@ -301,11 +314,7 @@ public final class SegmentedRaftLog extends RaftLogBase {
     // the entry is not in the segment's cache. Load the cache without holding the lock.
     getRaftLogMetrics().onRaftLogCacheMiss();
     cacheEviction.signal();
-    ReferenceCountedObject<LogEntryProto> loaded = segment.loadCache(record);
-    if (loaded != null) {
-      loaded.retain();
-    }
-    return loaded;
+    return segment.loadCache(record);
   }
```

Reply:
think I did. https://github.com/apache/ratis/pull/1045/commits/1414bd63fd10ea27aa9699c71bb9d7ebc8dbb803.

Maybe it was pushed after you'd reviewed, sorry.